src/microbe_stage/MembraneWaterRipple.cs

```diff
@@ -0,0 +1,572 @@
+        {
+            if (pastPositions != null)
+                pastPositions[i] = Vector2.Zero;
+            if (membranePositionHistory != null)
+                membranePositionHistory[i] = Vector3.Zero;
+        }
+
+        InitializeShaderParameters();
+        currentCamera = GetViewport()?.GetCamera3D();
+        isCameraPositionValid = false;
+        if (waterPlane != null)
+            waterPlane.Visible = false;
+    }
+
+    /// <summary>
+    ///   Initialize the effect with the parent membrane
+    /// </summary>
+    public void Initialize(Membrane? membrane)
+    {
+        if (membrane == null)
+            return;
+
+        parentMembrane = membrane;
+        lastPosition = parentMembrane.GlobalPosition;
+
+        // Initialize position history with current position
+        for (int i = 0; i < PositionHistorySize; ++i)
+        {
+            if (membranePositionHistory != null)
+                membranePositionHistory[i] = lastPosition;
+        }
+
+        isEffectEnabled = true;
+        if (waterPlane != null)
+            waterPlane.Visible = true;
+        UpdatePosition();
+    }
+
+    public override void _Process(double delta)
+    {
+        if (!EnableEffect || parentMembrane == null || waterPlane == null)
+            return;
+
+        // Updates camera
+        visibilityCheckTimer += (float)delta;
+        if (visibilityCheckTimer >= VisibilityCheckInterval)
+        {
+            visibilityCheckTimer = 0.0f;
+            UpdateCameraCache();
+            isCurrentlyVisible = IsMembraneVisible();
+
+            // Toggle water plane visibility
+            if (waterPlane.Visible != (isCurrentlyVisible && isEffectEnabled))
+            {
+                waterPlane.Visible = isCurrentlyVisible && isEffectEnabled;
+            }
+        }
+
+        if (!isCurrentlyVisible || !isEffectEnabled)
+            return;
+
+        UpdatePosition();
+
+        // Update time for animation based on LOD level and movement
+        float timeScale = 1.0f;
+        if (currentSpeed > 0.1f)
+        {
+            // Scale time update by current LOD level
+            switch (currentLodLevel)
+            {
+                case LodLevel.VeryLow:
+                    timeScale = 1.0f + currentSpeed * 0.05f;
+                    break;
+
+                case LodLevel.Low:
+                    timeScale = 1.0f + currentSpeed * 0.08f;
+                    break;
+
+                case LodLevel.Medium:
+                    timeScale = 1.0f + currentSpeed * 0.1f;
+                    break;
+
+                case LodLevel.High:
+                    timeScale = 1.0f + currentSpeed * 0.12f;
+                    break;
+            }
+        }
+
+        timeAccumulator += (float)delta * timeScale;
+        waterMaterial?.SetShaderParameter(timeOffsetParam, timeAccumulator);
+
+        // Update movement parameters (with reduced frequency for distant objects)
+        if (currentLodLevel == LodLevel.VeryLow)
```

Comment:
Does this actually help anything? As this is a parented node then I think the full transform matrix has to be calculated each frame basically anyway so I'd expect this extra checking to just eat up more performance than it has any chance of helping with.